django/forms/models.py

```diff
@@ -1310,8 +1309,8 @@ def has_changed(self, initial, data):
             data = []
         if len(initial) != len(data):
             return True
-        initial_set = set(force_text(value) for value in self.prepare_value(initial))
-        data_set = set(force_text(value) for value in data)
+        initial_set = set(str(value) for value in self.prepare_value(initial))
```

Comment:
use set comprehension (I guess it might be better to defer this task to a separate commit)